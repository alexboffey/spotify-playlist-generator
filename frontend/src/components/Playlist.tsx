import React, { useState } from "react";
import { Query } from "react-apollo";
import { message, List, Button, Empty } from "antd";

import {
  GENERATE_PLAYLIST_QUERY,
  IGeneratePlaylistQuery
} from "../graphql/generatePlaylist";
import Track from "./Track";
import Header from "./Header";
import SeedsModal from "./SeedsModal";
import { IArtist } from "../interfaces";
import SeedTags from "./SeedTags";
import ExportAsPlaylistModal from "./ExportAsPlaylistModal";
import { PlayingContext } from "./PlayingContext";

interface IProps {
  seeds: Array<IArtist>;
  setSeeds: (seeds: Array<IArtist>) => void;
}

const Playlist: React.FunctionComponent<IProps> = ({ seeds, setSeeds }) => {
  const [seedsModalVisible, setSeedsModalVisible] = useState<boolean>(false);
  const [exportModalVisible, setExportModalVisible] = useState<boolean>(false);
  const [isPlaying, setIsPlaying] = useState<any>(null);
  const formattedSeeds = seeds.map(({ id }) => id).join(",");

  return (
    <Query<IGeneratePlaylistQuery>
      query={GENERATE_PLAYLIST_QUERY}
      variables={{ seeds: formattedSeeds }}
      notifyOnNetworkStatusChange
    >
      {({ data, loading, error, refetch }) => {
        if (error) {
          message.error(error);
          return <Empty description="Something went wrong..." />;
        }

        return (
          <React.Fragment>
            <Header title="Playlist" />
            <div
              className="playlist-controls"
              style={{
                display: "flex",
                alignItems: "center",
                marginBottom: "1.5rem"
              }}
            >
              <div
                className="playlist-controls__seeds"
                style={{
                  display: "flex",
                  alignItems: "center",
                  marginRight: "auto"
                }}
              >
                <Button
                  icon="plus"
                  onClick={() => setSeedsModalVisible(!seedsModalVisible)}
                  style={{ marginRight: "1rem" }}
                  disabled={loading}
                  title="Add Seeds"
                >
                  Add Seeds
                </Button>
                <SeedsModal
                  visible={seedsModalVisible}
                  setVisible={setSeedsModalVisible}
                  seeds={seeds}
                  setSeeds={setSeeds}
                  refetch={refetch}
                />

                <SeedTags seeds={seeds} setSeeds={setSeeds} refetch={refetch} />
              </div>

              <Button
                type="ghost"
                onClick={() => refetch()}
                icon="reload"
                disabled={loading}
                style={{ marginRight: ".5rem" }}
                title="Regenerate Playlist"
              />

              <Button
                type="ghost"
                icon="save"
                disabled={loading}
                style={{ marginRight: ".5rem" }}
                title="Export as Spotify Playlist"
                onClick={() => setExportModalVisible(!exportModalVisible)}
              />
              <ExportAsPlaylistModal
                visible={exportModalVisible}
                setVisible={setExportModalVisible}
                tracks={
                  data && !loading
                    ? data.generatePlaylist.tracks.map(({ uri }) => uri)
                    : []
                }
              />
            </div>

            <List loading={loading}>
              <PlayingContext.Provider
                value={{
                  currentlyPlaying: isPlaying,
                  setCurrentlyPlaying: val => {
                    setIsPlaying(val);
                  }
                }}
              >
                {data &&
                  !loading &&
                  data.generatePlaylist.tracks.map(track => (
                    <Track key={track.id} {...track} />
                  ))}
              </PlayingContext.Provider>
            </List>
          </React.Fragment>
        );
      }}
    </Query>
  );
};

export default Playlist;
