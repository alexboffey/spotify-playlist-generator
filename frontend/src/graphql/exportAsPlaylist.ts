import gql from "graphql-tag";

export const EXPORT_AS_PLAYLIST_MUTATION = gql`
  mutation EXPORT_AS_PLAYLIST_MUTATION(
    $playlistName: String
    $tracks: [String]!
  ) {
    exportAsPlaylist(playlistName: $playlistName, tracks: $tracks) {
      name
    }
  }
`;

export interface IExportAsPlaylistMutation {
  exportAsPlaylist: {
    name: string;
  };
}
