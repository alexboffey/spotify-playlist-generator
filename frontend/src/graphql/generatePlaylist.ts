import gql from "graphql-tag";

export const GENERATE_PLAYLIST_QUERY = gql`
  query GENERATE_PLAYLIST_QUERY($seeds: String) {
    generatePlaylist(seeds: $seeds) {
      tracks {
        id
        name
        preview_url
        artists {
          id
          name
        }
        album {
          id
          name
          images {
            url
          }
        }
      }
    }
  }
`;

export interface IGeneratePlaylistQuery {
  generatePlaylist: {
    tracks: Array<{
      id: string;
      name: string;
      preview_url?: string;
      artists: Array<{ id: string; name: string }>;
      album: { id: string; name: string; images: Array<{ url: string }> };
    }>;
  };
}
