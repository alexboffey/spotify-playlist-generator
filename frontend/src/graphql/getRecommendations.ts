import gql from "graphql-tag";

export const GET_RECOMMENDATIONS_QUERY = gql`
  query GET_RECOMMENDATIONS_QUERY($seed_artists: String!) {
    getRecommendations(seed_artists: $seed_artists) {
      tracks {
        id
        name
        artists {
          id
          name
        }
        album {
          id
          name
        }
      }
    }
  }
`;

export interface IGetRecommendationsQuery {
  getRecommendations: {
    tracks: Array<{
      id: string;
      name: string;
      artists: { id: string; name: string };
      album: { id: string; name: string };
    }>;
  };
}
