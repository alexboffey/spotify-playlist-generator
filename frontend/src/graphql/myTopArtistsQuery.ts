import gql from "graphql-tag";

export const MY_TOP_ARTISTS_QUERY = gql`
  query MY_TOP_ARTISTS_QUERY(
    $limit: Int = 20
    $time_range: TimeRange = long_term
  ) {
    myTopArtists(limit: $limit, time_range: $time_range) {
      items {
        id
      }
    }
  }
`;

export interface IMyTopArtistsQuery {
  myTopArtists: {
    items: Array<{ id: string }>;
  };
}
