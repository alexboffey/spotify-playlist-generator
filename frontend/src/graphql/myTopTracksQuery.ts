import gql from "graphql-tag";

export const MY_TOP_TRACKS_QUERY = gql`
  query {
    myTopTracks(limit: 5, time_range: short_term) {
      items {
        name
        id
        href
        album {
          name
        }
      }
      next
      previous
      total
      limit
      href
    }
  }
`;

export interface MyTopTracksData {
  myTopTracks: {
    items: Array<{ name: string; id: string; href: string }>;
    next: string;
    previous: string;
    total: number;
    limit: number;
    href: string;
  };
}
