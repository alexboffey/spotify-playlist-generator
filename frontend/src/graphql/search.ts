import gql from "graphql-tag";

export const SEARCH_ARTISTS_QUERY = gql`
  query SEARCH_ARTISTS_QUERY(
    $query: String!
    $type: SearchType = artist
    $limit: Int
  ) {
    search(query: $query, type: $type, limit: $limit) {
      artists {
        items {
          name
        }
      }
    }
  }
`;

export const SEARCH_TRACKS_QUERY = gql`
  query SEARCH_ARTISTS_QUERY(
    $query: String!
    $type: SearchType = track
    $limit: Int
  ) {
    search(query: $query, type: $type, limit: $limit) {
      tracks {
        items {
          name
        }
      }
    }
  }
`;

export interface ISearchQuery {
  search: string;
  type: string;
}
