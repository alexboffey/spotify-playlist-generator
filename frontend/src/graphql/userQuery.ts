import gql from "graphql-tag"

export const USER_QUERY = gql`
  query {
    me {
      id
      name
      email
      spotifyId
    }
  }
`;

export interface IUserQuery {
  me: {
    id: string;
    name: string;
    email: string;
    spotifyId: string;
  };
}
