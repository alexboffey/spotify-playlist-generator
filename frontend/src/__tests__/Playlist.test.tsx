import React from "react";
import { mount } from "enzyme";
import toJSON from "enzyme-to-json";
import wait from "waait";
import { MockedProvider } from "react-apollo/test-utils";

import { GENERATE_PLAYLIST_QUERY } from "../graphql/generatePlaylist";
import Playlist from "../components/Playlist";

const mocks = [
  {
    request: {
      query: GENERATE_PLAYLIST_QUERY,
      variables: {
        seeds: ""
      }
    },
    result: {
      data: {
        generatePlaylist: {
          __typename: "test",
          tracks: [
            {
              __typename: "test",
              id: "trackId",
              name: "trackName",
              preview_url: "trackPreview",
              uri: "trackUri",
              artists: [
                {
                  id: "artistId",
                  name: "artistName",
                  __typename: "test"
                }
              ],
              album: {
                id: "albumId",
                name: "albumName",
                images: [
                  {
                    url: "albumeImageUrl",
                    __typename: "test"
                  }
                ],
                __typename: "test"
              }
            }
          ],
          seeds: [{ id: "seedId", type: "seedType", __typename: "test" }]
        }
      }
    }
  },
  {
    request: {
      query: GENERATE_PLAYLIST_QUERY,
      variables: {
        seeds: ""
      }
    },
    result: {
      data: {
        generatePlaylist: {
          __typename: "test",
          tracks: [
            {
              __typename: "test",
              id: "trackId",
              name: "trackName",
              preview_url: "trackPreview",
              uri: "trackUri",
              artists: [
                {
                  id: "artistId",
                  name: "artistName",
                  __typename: "test"
                }
              ],
              album: {
                id: "albumId",
                name: "albumName",
                images: [
                  {
                    url: "albumeImageUrl",
                    __typename: "test"
                  }
                ],
                __typename: "test"
              }
            },
            {
              __typename: "test",
              id: "trackId2",
              name: "trackName",
              preview_url: "trackPreview",
              uri: "trackUri",
              artists: [
                {
                  id: "artistId2",
                  name: "artistName",
                  __typename: "test"
                }
              ],
              album: {
                id: "albumId2",
                name: "albumName",
                images: [
                  {
                    url: "albumeImageUrl",
                    __typename: "test"
                  }
                ],
                __typename: "test"
              }
            }
          ],
          seeds: [{ id: "seedId", type: "seedType", __typename: "test" }]
        }
      }
    }
  }
];

describe("<Playlist />", () => {
  const setSeeds = jest.fn();

  beforeEach(() => {
    jest.resetAllMocks();
  });

  it("Queries, renders data and matches snapshot", async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Playlist seeds={[]} setSeeds={setSeeds} />
      </MockedProvider>
    );

    await wait();
    wrapper.update();

    expect(toJSON(wrapper.find(".ant-list"))).toMatchSnapshot();
  });

  it("Refetches data with load state after regenerate button is pressed", async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Playlist seeds={[]} setSeeds={setSeeds} />
      </MockedProvider>
    );

    await wait();
    wrapper.update();

    expect(wrapper.find(".ant-list-item")).toHaveLength(1);

    // Simulate event
    wrapper
      .find("button.ant-btn[title='Regenerate Playlist']")
      .simulate("click");

    expect(wrapper.find(".ant-spin-nested-loading")).toHaveLength(1);

    // Wait for update again
    await wait();
    wrapper.update();

    expect(wrapper.find(".ant-list-item")).toHaveLength(2);
    expect(toJSON(wrapper.find(".ant-list"))).toMatchSnapshot();
  });
});
