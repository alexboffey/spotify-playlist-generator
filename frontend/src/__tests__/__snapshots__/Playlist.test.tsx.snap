// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Playlist /> Queries, renders data and matches snapshot 1`] = `
<div
  className="ant-list ant-list-split"
>
  <Spin
    size="default"
    spinning={false}
    wrapperClassName=""
  >
    <div
      className="ant-spin-nested-loading"
    >
      <div
        className="ant-spin-container"
        key="container"
      >
        <Track
          __typename="test"
          album={
            Object {
              "__typename": "test",
              "id": "albumId",
              "images": Array [
                Object {
                  "__typename": "test",
                  "url": "albumeImageUrl",
                },
              ],
              "name": "albumName",
            }
          }
          artists={
            Array [
              Object {
                "__typename": "test",
                "id": "artistId",
                "name": "artistName",
              },
            ]
          }
          id="trackId"
          key="trackId"
          name="trackName"
          preview_url="trackPreview"
          uri="trackUri"
        >
          <Item>
            <li
              className="ant-list-item"
            >
              <Meta
                avatar={
                  <Avatar
                    shape="square"
                    size="large"
                    src="albumeImageUrl"
                  >
                    t
                  </Avatar>
                }
                description="artistName"
                title={
                  <span>
                    trackName
                  </span>
                }
              >
                <div
                  className="ant-list-item-meta"
                >
                  <div
                    className="ant-list-item-meta-avatar"
                  >
                    <Avatar
                      shape="square"
                      size="large"
                      src="albumeImageUrl"
                    >
                      <span
                        className="ant-avatar ant-avatar-lg ant-avatar-square ant-avatar-image"
                        style={Object {}}
                      >
                        <img
                          onError={[Function]}
                          src="albumeImageUrl"
                        />
                      </span>
                    </Avatar>
                  </div>
                  <div
                    className="ant-list-item-meta-content"
                  >
                    <h4
                      className="ant-list-item-meta-title"
                    >
                      <span>
                        trackName
                      </span>
                    </h4>
                    <div
                      className="ant-list-item-meta-description"
                    >
                      artistName
                    </div>
                  </div>
                </div>
              </Meta>
              <Button
                block={false}
                disabled={false}
                ghost={false}
                htmlType="button"
                icon="play-circle"
                loading={false}
                onClick={[Function]}
                title="Preview track"
              >
                <Wave>
                  <button
                    className="ant-btn"
                    disabled={false}
                    onClick={[Function]}
                    title="Preview track"
                    type="button"
                  >
                    <Icon
                      type="play-circle"
                    >
                      <LocaleReceiver
                        componentName="Icon"
                      >
                        <i
                          aria-label="icon: play-circle"
                          className="anticon anticon-play-circle"
                        >
                          <IconReact
                            className=""
                            type="play-circle-o"
                          >
                            <svg
                              aria-hidden="true"
                              className=""
                              data-icon="play-circle"
                              fill="currentColor"
                              focusable="false"
                              height="1em"
                              key="svg-play-circle"
                              viewBox="64 64 896 896"
                              width="1em"
                            >
                              <path
                                d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                                key="svg-play-circle-svg-0"
                              />
                              <path
                                d="M719.4 499.1l-296.1-215A15.9 15.9 0 0 0 398 297v430c0 13.1 14.8 20.5 25.3 12.9l296.1-215a15.9 15.9 0 0 0 0-25.8zm-257.6 134V390.9L628.5 512 461.8 633.1z"
                                key="svg-play-circle-svg-1"
                              />
                            </svg>
                          </IconReact>
                        </i>
                      </LocaleReceiver>
                    </Icon>
                    <span
                      key=".0"
                    >
                      Play
                    </span>
                  </button>
                </Wave>
              </Button>
            </li>
          </Item>
        </Track>
      </div>
    </div>
  </Spin>
</div>
`;

exports[`<Playlist /> Refetches data with load state after regenerate button is pressed 1`] = `
<div
  className="ant-list ant-list-split"
>
  <Spin
    size="default"
    spinning={false}
    wrapperClassName=""
  >
    <div
      className="ant-spin-nested-loading"
    >
      <div
        className="ant-spin-container"
        key="container"
      >
        <Track
          __typename="test"
          album={
            Object {
              "__typename": "test",
              "id": "albumId",
              "images": Array [
                Object {
                  "__typename": "test",
                  "url": "albumeImageUrl",
                },
              ],
              "name": "albumName",
            }
          }
          artists={
            Array [
              Object {
                "__typename": "test",
                "id": "artistId",
                "name": "artistName",
              },
            ]
          }
          id="trackId"
          key="trackId"
          name="trackName"
          preview_url="trackPreview"
          uri="trackUri"
        >
          <Item>
            <li
              className="ant-list-item"
            >
              <Meta
                avatar={
                  <Avatar
                    shape="square"
                    size="large"
                    src="albumeImageUrl"
                  >
                    t
                  </Avatar>
                }
                description="artistName"
                title={
                  <span>
                    trackName
                  </span>
                }
              >
                <div
                  className="ant-list-item-meta"
                >
                  <div
                    className="ant-list-item-meta-avatar"
                  >
                    <Avatar
                      shape="square"
                      size="large"
                      src="albumeImageUrl"
                    >
                      <span
                        className="ant-avatar ant-avatar-lg ant-avatar-square ant-avatar-image"
                        style={Object {}}
                      >
                        <img
                          onError={[Function]}
                          src="albumeImageUrl"
                        />
                      </span>
                    </Avatar>
                  </div>
                  <div
                    className="ant-list-item-meta-content"
                  >
                    <h4
                      className="ant-list-item-meta-title"
                    >
                      <span>
                        trackName
                      </span>
                    </h4>
                    <div
                      className="ant-list-item-meta-description"
                    >
                      artistName
                    </div>
                  </div>
                </div>
              </Meta>
              <Button
                block={false}
                disabled={false}
                ghost={false}
                htmlType="button"
                icon="play-circle"
                loading={false}
                onClick={[Function]}
                title="Preview track"
              >
                <Wave>
                  <button
                    className="ant-btn"
                    disabled={false}
                    onClick={[Function]}
                    title="Preview track"
                    type="button"
                  >
                    <Icon
                      type="play-circle"
                    >
                      <LocaleReceiver
                        componentName="Icon"
                      >
                        <i
                          aria-label="icon: play-circle"
                          className="anticon anticon-play-circle"
                        >
                          <IconReact
                            className=""
                            type="play-circle-o"
                          >
                            <svg
                              aria-hidden="true"
                              className=""
                              data-icon="play-circle"
                              fill="currentColor"
                              focusable="false"
                              height="1em"
                              key="svg-play-circle"
                              viewBox="64 64 896 896"
                              width="1em"
                            >
                              <path
                                d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                                key="svg-play-circle-svg-0"
                              />
                              <path
                                d="M719.4 499.1l-296.1-215A15.9 15.9 0 0 0 398 297v430c0 13.1 14.8 20.5 25.3 12.9l296.1-215a15.9 15.9 0 0 0 0-25.8zm-257.6 134V390.9L628.5 512 461.8 633.1z"
                                key="svg-play-circle-svg-1"
                              />
                            </svg>
                          </IconReact>
                        </i>
                      </LocaleReceiver>
                    </Icon>
                    <span
                      key=".0"
                    >
                      Play
                    </span>
                  </button>
                </Wave>
              </Button>
            </li>
          </Item>
        </Track>
        <Track
          __typename="test"
          album={
            Object {
              "__typename": "test",
              "id": "albumId2",
              "images": Array [
                Object {
                  "__typename": "test",
                  "url": "albumeImageUrl",
                },
              ],
              "name": "albumName",
            }
          }
          artists={
            Array [
              Object {
                "__typename": "test",
                "id": "artistId2",
                "name": "artistName",
              },
            ]
          }
          id="trackId2"
          key="trackId2"
          name="trackName"
          preview_url="trackPreview"
          uri="trackUri"
        >
          <Item>
            <li
              className="ant-list-item"
            >
              <Meta
                avatar={
                  <Avatar
                    shape="square"
                    size="large"
                    src="albumeImageUrl"
                  >
                    t
                  </Avatar>
                }
                description="artistName"
                title={
                  <span>
                    trackName
                  </span>
                }
              >
                <div
                  className="ant-list-item-meta"
                >
                  <div
                    className="ant-list-item-meta-avatar"
                  >
                    <Avatar
                      shape="square"
                      size="large"
                      src="albumeImageUrl"
                    >
                      <span
                        className="ant-avatar ant-avatar-lg ant-avatar-square ant-avatar-image"
                        style={Object {}}
                      >
                        <img
                          onError={[Function]}
                          src="albumeImageUrl"
                        />
                      </span>
                    </Avatar>
                  </div>
                  <div
                    className="ant-list-item-meta-content"
                  >
                    <h4
                      className="ant-list-item-meta-title"
                    >
                      <span>
                        trackName
                      </span>
                    </h4>
                    <div
                      className="ant-list-item-meta-description"
                    >
                      artistName
                    </div>
                  </div>
                </div>
              </Meta>
              <Button
                block={false}
                disabled={false}
                ghost={false}
                htmlType="button"
                icon="play-circle"
                loading={false}
                onClick={[Function]}
                title="Preview track"
              >
                <Wave>
                  <button
                    className="ant-btn"
                    disabled={false}
                    onClick={[Function]}
                    title="Preview track"
                    type="button"
                  >
                    <Icon
                      type="play-circle"
                    >
                      <LocaleReceiver
                        componentName="Icon"
                      >
                        <i
                          aria-label="icon: play-circle"
                          className="anticon anticon-play-circle"
                        >
                          <IconReact
                            className=""
                            type="play-circle-o"
                          >
                            <svg
                              aria-hidden="true"
                              className=""
                              data-icon="play-circle"
                              fill="currentColor"
                              focusable="false"
                              height="1em"
                              key="svg-play-circle"
                              viewBox="64 64 896 896"
                              width="1em"
                            >
                              <path
                                d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                                key="svg-play-circle-svg-0"
                              />
                              <path
                                d="M719.4 499.1l-296.1-215A15.9 15.9 0 0 0 398 297v430c0 13.1 14.8 20.5 25.3 12.9l296.1-215a15.9 15.9 0 0 0 0-25.8zm-257.6 134V390.9L628.5 512 461.8 633.1z"
                                key="svg-play-circle-svg-1"
                              />
                            </svg>
                          </IconReact>
                        </i>
                      </LocaleReceiver>
                    </Icon>
                    <span
                      key=".0"
                    >
                      Play
                    </span>
                  </button>
                </Wave>
              </Button>
            </li>
          </Item>
        </Track>
      </div>
    </div>
  </Spin>
</div>
`;
